### PR TITLE
Build improvements

### PR DESCRIPTION
**Pull request by jelz**
This PR:

- [x] Cuts build 2-3 times thanks to use of concurrency=2 (matching number of vCPUs) for all tasks but bundling (experienced hanging processes), use of `npm ci`, skipping code coverage instrumentation, not reinstalling dependencies in deployment step
- [x] Separates building artifacts from testing/deploying them
- [x] Stops to default deployment stage to production (to avoid deployment to production from local environments)
- [x] Always deploys all apps to verify that the process wasn't broken in a feature branch

Note that the test deployment from all branches ends up in the same location (same bucket, same stage). It's not intended to be used for sharing because builds will eventually override each other. The sole purpose of deployment is to test if it wasn't broken.

Screen Shot 2020-02-14 at 15 01 03

### CODE REVIEW

File: apps/smartling/lambda/package.json [hunk 1]
```diff
@@ -4,17 +4,12 @@
...
-    "build-lambda": "npm run clean && tsc",
-    "build-frontend": "cd ../frontend && npm run build",
-    "build": "npm run clean && npm run build-lambda && npm run build-frontend",
+    "build": "rimraf built && tsc",
     "lint": "tslint --project ./tsconfig.json",
-    "lint-fix": "tslint --project ./tsconfig.json --fix",
     "pretty": "prettier --config ./.prettierrc --write '{src/*.ts,src/*.spec.ts}'",
-    "test": "LOCAL_DEV=true jest **/*.spec.ts",
-    "test:ci": "npm run build-frontend && npm run test",
-    "test-dev": "LOCAL_DEV=true jest **/*.spec.ts --watch",
```
Comment (davidfateh) [inline apps/smartling/lambda/package.json]: `LOCAL_DEV=true` may be needed for local testing. A few things have changed here so maybe it is outdated.

Comment (jelz) [reply apps/smartling/lambda/package.json]: We only use it here: https://github.com/contentful/apps/blob/master/apps/smartling/lambda/src/index.ts#L119-L122 to treat app file as a server.

Tests still work w/o that.

File: .circleci/config.yml [hunk 2]
```diff
@@ -1,54 +1,56 @@
...
+      - run:
+          name: Build
+          command: npm run build
       - run:
           name: Test
           command: npm run test
+      - run:
+          name: Configure deployment
+          command: |
+            static_s3_base="s3://cf-apps-static/apps-test-$CIRCLE_SHA1"
```
Comment (lehnerm) [inline .circleci/config.yml]: Let's please not use the prd bucket for test deployments - we should get a second one for that.

File: .circleci/config.yml [hunk 3]
```diff
@@ -1,54 +1,56 @@
...
-      - image: circleci/node:10-stretch
-        environment:
-          STATIC_APPS_BUCKET: s3://cf-apps-static
-    steps:
-      - checkout
+            static_jira_s3_base="s3://cf-apps-static/apps-test-$CIRCLE_SHA1/jira"
+            test "$CIRCLE_BRANCH" == 'master' && static_jira_s3_base='s3://cf-apps-jira'
+            echo "export STATIC_JIRA_S3_BASE=\"$static_jira_s3_base\"" >> $BASH_ENV
+
+            stage='test'
```
Comment (lehnerm) [inline .circleci/config.yml]: I disagree with the fact that we need to have lambda deployment run on every branch. Especially with this setup, in which weird  side effects will most certainly occur when deployment is run on a single stage from multiple branches.

File: package.json [hunk 4]
```diff
@@ -1,14 +1,15 @@
...
   "name": "apps",
+  "private": true,
   "scripts": {
-    "lint": "lerna run lint --concurrency=1",
-    "test": "lerna run test:ci --concurrency=1",
-    "bootstrap": "lerna bootstrap --no-ci --concurrency=1",
-    "deploy": "lerna run deploy",
-    "postinstall": "npm run bootstrap"
+    "bootstrap": "lerna bootstrap --no-ci",
+    "bootstrap:ci": "lerna bootstrap --ci --concurrency=2",
```
Comment (lehnerm) [inline package.json]: I've disabled lerna concurrency due to CI flakiness for both test and bootstrap. Couldn't find much information on that - any experience with this?

File: .circleci/config.yml (hunk 3 above)
Comment (jelz) [reply .circleci/config.yml]: I think benefits outweigh problems. Right now the first time we attempt to deploy from CI to Lambda is after we merge to master. This blocks further deployments until the problem is solved (likely: CI role cannot do something) or the change is reverted. By attempting to deploy in branches we mitigate this risk.

File: package.json (hunk 4 above)
Comment (jelz) [reply package.json]: So I did some digging and:

- Using `npm ci` (so `lerna --ci`) is faster and more lightweight (according to npm: It can be significantly faster than a regular npm install by skipping certain user-oriented features) - never experienced any crashes/hanging with `--ci`
- Linting, testing and deployment can also be reliably run with concurrency of 2. 2 matches number of vCPUs we get in Circle
- The only problematic step is the build step. I don't know if it's problem with parcel (cache interference?) or it's so CPU-bound that it has hard time running in parallel but this is the only step that requires concurrency of 1 for stability

In the previous model building artifacts was part of test and deploy steps (sometimes repeated). Now we build all packages and then run tests on them and deploy in separate steps.

To sum up we build in series and everything else is done with parallelism of 2.

File: .circleci/config.yml (hunk 2 above)
Comment (jelz) [reply .circleci/config.yml]: Makes sense.

File: .circleci/config.yml (hunk 3 above)
Comment (lehnerm) [reply .circleci/config.yml]: As long as we just block master, I think this is fine. In case we need to ship something else fast, we can still revert the changes. If we're worried that the prd deployment might actually break something important, I'd suggest we add a dev deployment stage in the master build.

Comment (jelz) [reply .circleci/config.yml]: Do you mean merge to master = staging, merge to `prod` branch = deployment to production?

Comment (jelz) [reply .circleci/config.yml]: Other issue with the current state of affairs: frontend gets deployed, backend deployment fails. Frontend relies on new backend version. App is broken.

Comment (lehnerm) [reply .circleci/config.yml]: I mean that the master build would be something like
1) deploy to `dev`
2) smoke test dev (:smiling_imp:)
3) deploy to `prd`

Frontend and backend will always have to remain compatible between individual changes, since we cannot guarantee an instant client side code update due to caching anyway. We also cannot gurantee the order of deployment.

I'd also like to avoid the `prod` branch approach. I think we both agreed that it is a very problematic solution.

File: package.json (hunk 4 above)
Comment (lehnerm) [reply package.json]: I was running into this with bootstrap: https://github.com/lerna/lerna/issues/1457
